### PR TITLE
Fix for KV_V2 Custom Metadata Bug

### PR DESCRIPTION
**Pull request by AnPucel**
This handles the case in which users are using the new `KV_V2` client library with an older version of Vault (~1.8) that doesn't have an implementation for `custom_metadata`. Currently this use case is failing when attempting to read. 

#16870

### CODE REVIEW

File: api/kv_test.go [hunk 1]
```diff
@@ -319,6 +319,22 @@ func TestExtractCustomMetadata(t *testing.T) {
 			},
 			expected: map[string]interface{}{"org": "eng"},
 		},
+		{
+			name: "a read response with no custom metadata",
```
Comment (digivava) [inline api/kv_test.go:323]: I think these will work great. I wonder if it's also worth adding a test case for a read and write response that has the `custom_metadata` field but where it's just an empty map. Just to make sure this function works when there _is_ a custom_metadata field but the custom metadata that was in there before got deleted (cuz I think it would end up as an empty map in that case, not nil. Not 100% sure but yeah might as well test?)

Comment (AnPucel) [reply api/kv_test.go:323]: Good call. I can add that test as well.

**Review by digivava — APPROVED** [review]
This should work now! Remember to release both the `api` module and the `api/auth` submodules with this change.

File: api/kv_v2.go [hunk 2]
```diff
@@ -497,30 +485,24 @@ func (kv *KVv2) Rollback(ctx context.Context, secretPath string, toVersion int)
...
+			return nil
 		}
+		customMetadataInterface = metadataMap["custom_metadata"]
 	}
 
 	cm, ok := customMetadataInterface.(map[string]interface{})
-	if !ok && customMetadataInterface != nil {
-		return nil, fmt.Errorf("unexpected type for 'metadata' element: %T (%#v)", customMetadataInterface, customMetadataInterface)
+	if !ok {
+		return nil
```
Comment (dhuckins) [inline api/kv_v2.go:503]: do we want to return nil or an empty map?

**Review by dhuckins — APPROVED** [review]

File: api/kv_v2.go [hunk 3]
```diff
@@ -497,30 +485,24 @@ func (kv *KVv2) Rollback(ctx context.Context, secretPath string, toVersion int)
 	return kvs, nil
 }
 
-func extractCustomMetadata(secret *Secret) (map[string]interface{}, error) {
+func extractCustomMetadata(secret *Secret) map[string]interface{} {
 	// Logical Writes return the metadata directly, Reads return it nested inside the "metadata" key
 	customMetadataInterface, ok := secret.Data["custom_metadata"]
+
```
Comment (averche) [inline api/kv_v2.go:491]:
```suggestion
```

File: api/kv_v2.go [hunk 4]
```diff
@@ -497,30 +485,24 @@ func (kv *KVv2) Rollback(ctx context.Context, secretPath string, toVersion int)
...
 
 	cm, ok := customMetadataInterface.(map[string]interface{})
-	if !ok && customMetadataInterface != nil {
-		return nil, fmt.Errorf("unexpected type for 'metadata' element: %T (%#v)", customMetadataInterface, customMetadataInterface)
+	if !ok {
+		return nil
 	}
-
-	return cm, nil
+	return cm
```
Comment (averche) [inline api/kv_v2.go:505]:
```suggestion

	return cm
```

**Review by averche — APPROVED** [review]
A couple style suggestions, otherwise LGTM 👍

File: api/kv_v2.go (hunk 2 above)
Comment (averche) [reply api/kv_v2.go:503]: Empty map might be more convenient for the end users but `nil` is more common to check against 🤷 I guess I would be OK with either.

File: api/kv_v2.go (hunk 3 above)
Comment (AnPucel) [reply api/kv_v2.go:491]: It's the C# in me that keeps adding random lines.

File: api/kv_v2.go (hunk 2 above)
Comment (digivava) [reply api/kv_v2.go:503]: I think representing it similar to the original API response would be good, where I believe an unset custom_metadata map is `nil` and a set-at-one-point-but-now-all-keys-have-been-deleted custom_metadata map is an empty map.
